### ISSUE TITLE
reg_f3d crash with long inputs

### ISSUE THREAD

**Issue by mmodat** [KCL-BMEIS/niftyreg#13]
In gitlab by @imalone on Feb 14, 2017, 15:46

I hit a hard-to-trace crash with reg_f3d, which appeared to complete okay, but triggered a general protection fault on completion and reported large negative "Performed in" times. This would cause niftypipe to halt (in any case outputs would have been suspect). It turns out the buffer used for text to report the input command line was being overrun. Eventually tracked it down with -fsanitize=address.

I'd suggest using asprintf to fix this:
```
diff --git a/reg-apps/reg_f3d.cpp b/reg-apps/reg_f3d.cpp
index 42e9f59..787ff2e 100755
--- a/reg-apps/reg_f3d.cpp
+++ b/reg-apps/reg_f3d.cpp
@@ -173,7 +173,7 @@ int main(int argc, char **argv)
    omp_set_num_threads(defaultOpenMPValue);
 #endif
 
-   char text[1024];
+   char *text=0, *tmptext;
    //\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/
    // Check if any information is required
    for(int i=1; i<argc; i++)
@@ -218,11 +218,21 @@ int main(int argc, char **argv)
 #endif
       reg_print_info((argv[0]), "");
       reg_print_info((argv[0]), "Command line:");
-      sprintf(text, "\t");
-      for(int i=0; i<argc; i++)
-         sprintf(text+strlen(text), " %s", argv[i]);
+      if (asprintf(&text, "\t") < 0) {
+       reg_print_info((argv[0]), "Out of memory storing command line");
+       return EXIT_FAILURE;
+      }
+      for(int i=0; i<argc; i++) {
+       if (asprintf(&tmptext, "%s %s", text, argv[i]) < 0) {
+         reg_print_info((argv[0]), "Out of memory storing command line");
+         return EXIT_FAILURE;
+       }
+       free(text);
+       text=tmptext;
+      }
       reg_print_info((argv[0]), text);
       reg_print_info((argv[0]), "");
+      free(text); text=0;
 #ifdef NDEBUG
    }
 #endif
@@ -730,8 +740,9 @@ int main(int argc, char **argv)
    if(verbose)
    {
       int maxThreadNumber = omp_get_max_threads();
-      sprintf(text, "OpenMP is used with %i thread(s)", maxThreadNumber);
+      asprintf(&text, "OpenMP is used with %i thread(s)", maxThreadNumber);
       reg_print_info((argv[0]), text);
+      free(text); text=0;
    }
 #endif // _OPENMP
 
@@ -843,9 +854,10 @@ int main(int argc, char **argv)
       time(&end);
       int minutes=(int)floorf((end-start)/60.0f);
       int seconds=(int)(end-start - 60*minutes);
-      sprintf(text, "Registration performed in %i min %i sec", minutes, seconds);
+      asprintf(&text, "Registration performed in %i min %i sec", minutes, seconds);
       reg_print_info((argv[0]), text);
       reg_print_info((argv[0]), "Have a good day !");
+      free(text); text=0;
 #ifdef NDEBUG
    }
 #endif
```

Small problem in that asprintf is non-standard, GNU and BSD only, but implementation is trivial in C99 using vsnprintf, this is my version:
```
#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>

#ifndef _GNU_SOURCE

int vasprintf(char **strp, const char *fmt, va_list ap) {
  va_list ap2;
  va_copy(ap2, ap);
  int str_len, need_len;
  str_len = vsnprintf(0, 0, fmt, ap);
  need_len = str_len + 1;
  *strp = malloc(need_len);
  if (*strp) {
    vsnprintf(*strp, need_len, fmt, ap2);
  } else {
    *strp = 0;
    str_len = -1;
  }
  va_end(ap2);
  return str_len;
}

int asprintf(char **strp, const char *fmt, ...) {
  va_list ap;
  va_start(ap, fmt);
  int ret_val = vasprintf(strp, fmt, ap);
  va_end(ap);
  return ret_val;
}

#endif
```

**Comment by mmodat** [comment]
In gitlab by @imalone on Feb 14, 2017, 15:51

@mmodat I'm not sure if you automatically get notified of niftyreg issues.

**Comment by mmodat** [comment]
In gitlab by @imalone on Feb 15, 2017, 17:54

Yesterday's comment that didn't submit: "Actually, this being c++, could just use the string type instead."

Since the asprintf solution needs a bit more code to make it portable, I've re-done this with C++'s ostringstream, which should just work. (I noticed the buffer size has been upped a bit since the checkout I was working on, but especially as this is user facing, a dynamic approach might be better.) If I submit a branch for this can you point me towards some build instructions to check Windows and clang? (Don't have a Mac, but looks like I can install it on Linux.)

**Comment by mmodat** [comment]
In gitlab by @imalone on Feb 16, 2017, 14:16

@mmodat 
The correct stringstream solution is actually a bit ugly, you need a separate string variable as well,
```
std::ostringstream text_ss;
std::string text;
int m_int = 10;

text_ss << "Add some things, n=" << m_int;
text = text_ss.str();
function_needing_charptr (text.c_str);
// If need to re-use later
text_ss.clear() ; text_ss.str("");
```

While asprintf is nearly a drop-in replacement for the existing code:
```
char *text;
char *m_char = "some string";
char *m_int = 10;

asprintf(&text, "Some format %i, %s", m_int, m_char);
function_needing_charptr(text);
// Should really do if want to re-use later, but not compulsory
free(text);
```

Adding more things, ostringstream:
```
for (int ii = 0 ; ii<10; ii++){
  text << "N: " << ii << " ";
}
```

Adding more things, asprintf:
```
char *tmptext;
for (int ii = 0 ; ii<10; ii++){
  asprintf(&tmptext, "%s N:%i ", text, tmptext);
  free(text);
  text=tmptext;
}
```

Any preference? Providing an asprintf implementation to cover windows is fairly straightforward (the BSD one would be licensing compatible if we didn't want to use a home-rolled version). I'm not that keen on static buffers for variable length input, even if we don't think it's a security issue you run into it in normal use eventually and it's often hard to track the root cause of weirdness.

**Comment by mmodat** [comment]
In gitlab by @imalone on Feb 21, 2017, 13:43

Ran into another of these which is more deeply embedded, the reporting in _reg_f3d.cpp:reg_f3d<T>::Initialise() sprintf()s a file name, when the filenames get long in a pipeline they exceed the current limit, resulting in a hard to diagnose crash during the Initialise step. Suspect _reg_aladin.cpp does similar.

**Comment by mmodat** [comment]
In gitlab by @imalone on Feb 28, 2017, 16:46

@mmodat any thoughts on this? I've got reg_f3d working for this case by doing a minimal refactoring with asprintf on that reporting section. Most places in the code where a static buffer is used are okay because they only use numeric format specifiers, but the ones reporting filenames are vulnerable to overrunning.

**Comment by mmodat** [comment]
In gitlab by @imalone on Mar 13, 2017, 15:42

@mmodat 
I've come up with a neater solution, added a string function that does what asprintf does.

I'll see if I can work out how to build on VisualStudio, are you able to get someone to check [this branch](mmodat/niftyreg/tree/13-reg_f3d-crash-with-long-inputs) on Mac?

Changes required to use it are quite minimal:
```
-   char text[2048];
+   std::string text;
```

```
-      sprintf(text, "\t");
+      text = "\t";
```

```
-      sprintf(text, "OpenMP is used with %i thread(s)", maxThreadNumber);
-      reg_print_info((argv[0]), text);
+      text = stringFormat("OpenMP is used with %i thread(s)", maxThreadNumber);
+      reg_print_info((argv[0]), text.c_str());
```

No need to add any free() statements, and additional bonus, this works for appending:
```
text = stringFormat("%s yadayada", text.c_str());
```

So far I've updated the reporting parts of reg_f3d and reg_aladin that use filenames, and one function in reg-io/_reg_ReadWriteImage.cpp which may not really have needed it. It could be used to replace all the sprintf() I think, but not worth attempting till we know if it's a workable solution.

**Comment by mmodat** [comment]
In gitlab by @imalone on Mar 13, 2017, 18:23

Compiles okay in clang and gcc on Linux, reg_aladin and reg_f3d both run, reg_f3d is okay with the case that was crashing. Trying to set up a windows build environment currently...

**Comment by mmodat** [comment]
In gitlab by @imalone on Mar 15, 2017, 16:59

@mmodat 
I can't compile current HEAD in Visual Studio, even without my patches. Have been battling with getting the development environment for that set up for a few days now and going to have to give up. Do you know anyone with access to a Windows environment that will build niftyreg who could check these work? These fixes, and checking niftyfit builds are among the last few things I need to complete the diffusion pipeline.

**Comment by mmodat** [comment]
In gitlab by @mmodat on Mar 20, 2017, 12:39

I will check it on Windows

**Comment by mmodat** [comment]
In gitlab by @imalone on Mar 20, 2017, 13:06

Hi, I managed on Friday to build it on visual studio 2013 (oldest one I could find), unpatched it has the crash with long filenames, and patched it builds okay and runs with no crash. Can't guarantee about visual studio 2010 though (the stdarg va_copy it uses is C++ 11 or C99).

**Comment by mmodat** [comment]
In gitlab by @mmodat on May 11, 2017, 19:03

mentioned in merge request !10

**Comment by mmodat** [comment]
In gitlab by @mmodat on May 11, 2017, 19:05

closed via commit c289b3bfddddfe950fb658254ec10d5c41202a52
